[Official lesson](https://pages.opencodingsociety.com/python/developing-procedures/2026) · [All Python homework]({{site.baseurl}}/homework/)

Each Python runner is independent. Recorded outputs below are from local execution; random results and date-based reports can change when rerun.

## Exercise 1 — descriptive names

I renamed the broad make/mix/bake steps to prepare_vanilla_cupcakes, measure_dry_ingredients, combine_batter, and bake_cupcakes. They describe the purpose without opening every function.

In [1]:
# CODE_RUNNER: Exercise 1 — descriptive names

def measure_dry_ingredients():
    print("Measure 2 cups flour and 1 cup sugar")

def combine_batter():
    print("Mix dry ingredients with eggs and milk")

def bake_cupcakes():
    print("Bake cupcakes at 350 F for 20 minutes")

def prepare_vanilla_cupcakes():
    measure_dry_ingredients()
    combine_batter()
    bake_cupcakes()

prepare_vanilla_cupcakes()


Measure 2 cups flour and 1 cup sugar
Mix dry ingredients with eggs and milk
Bake cupcakes at 350 F for 20 minutes


### Recorded output

<pre>Measure 2 cups flour and 1 cup sugar
Mix dry ingredients with eggs and milk
Bake cupcakes at 350 F for 20 minutes</pre>

## Naming mini-game

Editable variables replace keyboard prompts so Run works on the website. Try prepare_dessert and do_step; both execute the same sequence, but only the first communicates the goal. Blank names use a default; cancellation skips execution.

In [2]:
# CODE_RUNNER: Naming mini-game

import random

def play_game(procedure_name, command):
    name = procedure_name.strip() or "prepare_ingredient"
    def prepare_dessert():
        ingredient = random.choice(["flour", "sugar", "eggs", "milk"])
        print(f"Mixing {ingredient} into the bowl")
        print("Heating the oven")
        print("Dessert is ready!")
    print("Procedure name:", name)
    if command.strip().lower() == "run":
        prepare_dessert()
        return True
    print("Canceled")
    return False

assert play_game("prepare_dessert", "run")
assert play_game("do_step", "run")
assert not play_game("", "cancel")


Procedure name: prepare_dessert
Mixing milk into the bowl
Heating the oven
Dessert is ready!
Procedure name: do_step
Mixing sugar into the bowl
Heating the oven
Dessert is ready!
Procedure name: prepare_ingredient
Canceled


### Recorded output

<pre>Procedure name: prepare_dessert
Mixing milk into the bowl
Heating the oven
Dessert is ready!
Procedure name: do_step
Mixing sugar into the bowl
Heating the oven
Dessert is ready!
Procedure name: prepare_ingredient
Canceled</pre>

## Exercise 2 — Java methods

The official page labels one JavaScript example as Java; the following is actual Java. The public method is the reusable entry point and the private helpers are implementation details accessible inside this class. `static` lets the example run without constructing an object.

<pre><code>public class DessertRobot {
    public static void prepareVanillaCupcakes() {
        measureDryIngredients();
        combineBatter();
        bakeCupcakes();
    }
    private static void measureDryIngredients() {
        System.out.println(&quot;Measure 2 cups flour and 1 cup sugar&quot;);
    }
    private static void combineBatter() {
        System.out.println(&quot;Mix dry ingredients with eggs and milk&quot;);
    }
    private static void bakeCupcakes() {
        System.out.println(&quot;Bake cupcakes at 350 F for 20 minutes&quot;);
    }
    public static void main(String[] args) {
        prepareVanillaCupcakes();
    }
}</code></pre>

## Exercise 3 — decomposition and tests

I split recipe planning into scaling, temperature conversion, and coordination. Each pure helper can be tested without running the whole recipe. Zero batches and an invalid negative count test boundaries.

In [3]:
# CODE_RUNNER: Exercise 3 — decomposition and tests

def scale_recipe(batches):
    """Return flour cups and sugar cups for a nonnegative number of batches."""
    if batches < 0:
        raise ValueError("Batches cannot be negative")
    return {"flour_cups": 2 * batches, "sugar_cups": batches}

def fahrenheit_to_celsius(fahrenheit):
    return (fahrenheit - 32) * 5 / 9

def plan_cupcakes(batches):
    return {"ingredients": scale_recipe(batches), "oven_celsius": round(fahrenheit_to_celsius(350), 1)}

assert scale_recipe(0) == {"flour_cups": 0, "sugar_cups": 0}
assert scale_recipe(2) == {"flour_cups": 4, "sugar_cups": 2}
assert fahrenheit_to_celsius(32) == 0
assert fahrenheit_to_celsius(212) == 100
try:
    scale_recipe(-1)
except ValueError as error:
    print("Negative test:", error)
else:
    raise AssertionError("Negative batches must fail")
print("Two-batch plan:", plan_cupcakes(2))
print("Helper and integration tests passed")


Negative test: Batches cannot be negative
Two-batch plan: {'ingredients': {'flour_cups': 4, 'sugar_cups': 2}, 'oven_celsius': 176.7}
Helper and integration tests passed


### Recorded output

<pre>Negative test: Batches cannot be negative
Two-batch plan: {&#x27;ingredients&#x27;: {&#x27;flour_cups&#x27;: 4, &#x27;sugar_cups&#x27;: 2}, &#x27;oven_celsius&#x27;: 176.7}
Helper and integration tests passed</pre>